File: 94.R
# Almost equilateral triangles

# Constants
MAX <- 1e9

total <- 0

m <- 1
while ( TRUE ) {
    m <- m + 1
    m2 <- m * m
    df <- data.frame(n=seq.int(m - 1))
    df$parity <- (m + df$n) %% 2
    df <- df[df$parity == 1,]
    df$n2 <- df$n * df$n
    df$a <- m2 - df$n2
    df$b <- 2 * m * df$n
    df$c <- m2 + df$n2
    df$absa <- abs(2 * df$a - df$c)
    df$absb <- abs(2 * df$b - df$c)
    df <- df[df$absa == 1 | df$absb == 1,]
    df$perima <- 2 * (df$a + df$c)
    df$perimb <- 2 * (df$b + df$c)
    if ( sum(df$perima > MAX & df$perimb > MAX) ) {
        break
    }
    if ( sum(df$absa == 1) ) {
        total <- total + df$perima[df$absa == 1]
    }
    if ( sum(df$absb == 1) ) {
        total <- total + df$perimb[df$absb == 1]
    }
}

cat(total, fill=TRUE)

File: 94.js
#!/usr/bin/env node

// Almost equilateral triangles

var MAX = 1e9;

var total = 0;

var m = 1;
var below_max = true;
while (below_max) {
    m++;
    var m2 = m * m;
    var n = m % 2 ? 0 : -1;
    while (n < m - 2) {
        n += 2;
        var n2 = n * n;
        var a = m2 - n2;
        var b = 2 * m * n;
        var c = m2 + n2;
        if (Math.abs(2 * a - c) != 1 && Math.abs(2 * b - c) != 1) {
            continue;
        }
        if (2 * (a + c) > MAX && 2 * (b + c) > MAX) {
            below_max = false;
            break;
        }
        if (Math.abs(2 * a - c) == 1) {
            total += 2 * (a + c);
        }
        if (Math.abs(2 * b - c) == 1) {
            total += 2 * (b + c);
        }
    }
}

console.log(total);
